### DISCUSSION TITLE
Open/closed universe

### DISCUSSION THREAD

**Discussion by MarcoMarinucci** [galacticusorg/galacticus#481] [Q&A] [answered]
Hello everyone,

I would run Galacticus with a cosmology which is not flat. 
Now, what I have always done was to run Gal. with the option <cosmologyFunctions value="matterLambda"/>, but I now put values for OmegaMatter and OmegaLambda such that OmegaMatter + OmegaLambda > or < 1 I have the following fatal error 

Fatal error:
no fitting formula available for this cosmology
 Occurred at:
     function:bryanNorman1998ConstructorInternal
         file:structure_formation.virial_density_contrast.Bryan_Norman.F90   [line 115]

What should I do? I wasn't able to find any other option for cosmologyFunctions in other scripts or in the notes. 

Thanks for helping!

Marco

**Comment by abensonca** [comment]
Hi Marco: The problem here is that [Bryan & Norman (1998)](https://ui.adsabs.harvard.edu/abs/1998ApJ...495...80B/abstract) only provide fitting functions for the virial density contrast $\Delta_\mathrm{vir}$ for flat universes, or open universe with $\Omega_\Lambda=0$. So, in your case it doesn't have a fitting function that can be applied.

The solution would be to use a different virial density contrast class. The `bryanNorman1998` class is just a fitting function to the results of the spherical collapse model. So, you could switch to the [`virialDensityContrastSphericalCollapseClsnlssMttrCsmlgclCnstnt`](https://github.com/galacticusorg/galacticus/releases/download/bleeding-edge/Galacticus_Development.pdf#virialDensityContrastSphericalCollapseClsnlssMttrCsmlgclCnstnt) class:
```
<virialDensityContrast value="sphericalCollapseClsnlssMttrCsmlgclCnstnt"/>
```
which implements the full spherical collapse model in cosmologies containing matter plus a cosmological constant, and should work for non-flat cases.

**Reply by MarcoMarinucci** [reply]
Hi Andrew, thanks.

I already have that option for the virial density contrast:


  Structure formation options 
  linearGrowth value="collisionlessMatter"
  criticalOverdensity value="sphericalCollapseClsnlssMttrCsmlgclCnstnt"
  virialDensityContrast value="sphericalCollapseClsnlssMttrCsmlgclCnstnt"

but I still have the error.
the cosmology I want to run is

   Cosmological parameters and options
  cosmologyFunctions value="matterLambda"
  cosmologyParameters value="simple"
    HubbleConstant value="66.62000"
    OmegaMatter value=" 0.3194"
    OmegaDarkEnergy value=" 0.7146"
   OmegaBaryon value=" 0.0502"                  
    temperatureCMB value=" 2.72548"
  cosmologyParameters


I also attach the xml file as a txt. Note that a similar file, but with OmegaMatter + OmegaLambda = 1 and exactly the same other parameters is normally running.

thanks!
Marco

[z1_sephigh_rs12.txt](https://github.com/galacticusorg/galacticus/files/12730782/z1_sephigh_rs12.txt)

**Reply by abensonca** [reply]
Ok, I see the problem. The model used by the `li2020` class for subhalo initial orbital parameters is based on the Bryan & Norman (1998) definition of virial density contrast. Galacticus internally converts to your preferred virial density contrast. But, in this case it can't do that as no fitting function is available.

Given that Bryan & Norman is a fit to the spherical collapse model I'll add an option to allow the `li2020` class to assume that no conversion is necessary. I'll let you know when I have this implemented.

**Comment by abensonca** [comment, marked as answer]
This should now be fixed in the [latest update](https://github.com/galacticusorg/galacticus/commit/456ee65559ae2d66240a0338a9207d59a749d28e) to Galacticus. For cosmologies where [Bryan & Norman (1998)](https://ui.adsabs.harvard.edu/abs/1998ApJ...495...80B/abstract) don't provide a fitting function it will automatically revert to using numerical solutions to the spherical top-hat collapse model.
